parser/prism/Translator.cc

```diff
@@ -1036,15 +1079,25 @@ std::unique_ptr<parser::Node> Translator::translateConstantPath(pm_constant_path
         parent = translate(node->parent);
     } else {                                                // This is a fully qualified constant reference, like `::A`.
         pm_location_t *delimiterLoc = &node->delimiter_loc; // The location of the `::`
-        parent = make_unique<parser::Cbase>(parser.translateLocation(delimiterLoc));
+        parent = make_node<parser::Cbase>(parser.translateLocation(delimiterLoc));
+        parent->cacheDesugaredExpr(MK::Constant(parent->loc, core::Symbols::root()));
     }
 
+    auto parentExpr = parent->getCachedDesugaredExpr();
+    ENFORCE(parentExpr != nullptr);
+
+    auto sorbetName = gs.enterNameConstant(name);
+
+    unique_ptr<parser::Node> sorbetNode;
     if (isAssignment) {
-        return make_unique<parser::ConstLhs>(parser.translateLocation(loc), std::move(parent),
-                                             gs.enterNameConstant(name));
+        sorbetNode = make_node<parser::ConstLhs>(parser.translateLocation(loc), std::move(parent), sorbetName);
     } else {
-        return make_unique<parser::Const>(parser.translateLocation(loc), std::move(parent), gs.enterNameConstant(name));
+        sorbetNode = make_node<parser::Const>(parser.translateLocation(loc), std::move(parent), sorbetName);
     }
+
+    sorbetNode->cacheDesugaredExpr(MK::UnresolvedConstant(sorbetNode->loc, std::move(parentExpr), sorbetName));
```

Comment:
This is an example where producing one Node's AST Expression requires consuming its children's AST Expressions. In this case, we can only make the AST Expression for the `C` in `::A::B::C` after we've made it for its child node that represents `::A::B`

```
   ::A::B:C  // <- Making this guy steals the `desugarExpr` out of the `::A::B` node below it.
     /     \
  ::A::B    C
    / \
  ::A  B
  /  \
CBase A
```